File: old/internal/log/log_manager.go
package log

import (
	"fmt"
	"kura/internal/file"
)

type LogManager struct {
	fileManager    *file.FileManager
	logFile        string
	logPage        *file.Page
	currentBlockID *file.BlockID
	latestLSN      int // latest log sequence number
	lastSavedLSN   int // last saved log sequence number
}

func NewLogManager(fileManager *file.FileManager, logFile string) (*LogManager, error) {
	logSize, err := fileManager.Length(logFile)
	if err != nil {
		return nil, fmt.Errorf("failed to get log file length %s: %w", logFile, err)
	}
	logPage := file.NewPage(fileManager.BlockSize())

	var currentBlockID *file.BlockID
	if logSize == 0 {
		newBlockID, err := appendNewBlock(fileManager, logFile, logPage)
		if err != nil {
			return nil, fmt.Errorf("failed to create new block for empty log file %s: %w", logFile, err)
		}
		currentBlockID = newBlockID
	} else {
		currentBlockID = file.NewBlockID(logFile, logSize-1)
		if err := fileManager.Read(currentBlockID, logPage); err != nil {
			return nil, fmt.Errorf("failed to read the last log block %s: %w", logFile, err)
		}
	}

	return &LogManager{
		fileManager:    fileManager,
		logFile:        logFile,
		logPage:        logPage,
		currentBlockID: currentBlockID,
		latestLSN:      0,
		lastSavedLSN:   0,
	}, nil
}

// Flush forces a log page to write to disk if the LSN exceeds last saved LSN.
func (lm *LogManager) Flush(lsn int) {
	if lsn >= lm.lastSavedLSN {
		lm.flush()
	}
}

func (lm *LogManager) Iterator() *LogIterator {
	lm.Flush(lm.latestLSN)
	return NewLogIterator(lm.fileManager, lm.currentBlockID)
}

// Append adds a new log record and returns the latest LSN.
func (lm *LogManager) Append(logRecord []byte) (int, error) {
	boundary := int(lm.logPage.GetInt(0))
	recordSize := len(logRecord)
	bytesNeeded := recordSize + file.Int32Size

	if boundary-bytesNeeded < file.Int32Size {
		lm.flush() // move to the next block
		newBlockID, err := appendNewBlock(lm.fileManager, lm.logFile, lm.logPage)
		if err != nil {
			return 0, fmt.Errorf("failed to append new block to log file %s: %w", lm.logFile, err)
		}
		lm.currentBlockID = newBlockID
		boundary = int(lm.logPage.GetInt(0))
	}

	recordPos := boundary - bytesNeeded
	lm.logPage.SetBytes(recordPos, logRecord)
	lm.logPage.SetInt(0, int32(recordPos))
	lm.latestLSN += 1
	return lm.latestLSN, nil
}

func appendNewBlock(fileManager *file.FileManager, logFile string, logPage *file.Page) (*file.BlockID, error) {
	blockID, err := fileManager.Append(logFile)
	if err != nil {
		return nil, fmt.Errorf("failed to append block to log file %s: %w", logFile, err)
	}
	logPage.SetInt(0, int32(fileManager.BlockSize()))
	if err := fileManager.Write(blockID, logPage); err != nil {
		return nil, fmt.Errorf("failed to write initial block data to log file %s: %w", logFile, err)
	}
	return blockID, nil
}

// flush writes the current log page to disk and updates the last saved LSN.
func (lm *LogManager) flush() {
	if err := lm.fileManager.Write(lm.currentBlockID, lm.logPage); err != nil {
		fmt.Printf("error flushing log page to disk for file %s: %v\n", lm.logFile, err)
		return
	}
	lm.lastSavedLSN = lm.latestLSN
}

func (lm *LogManager) CurrentBlockID() *file.BlockID {
	return lm.currentBlockID
}

File: app/src/main/kotlin/com/uesho/kura/log/LogManager.kt
package com.uesho.kura.log

import com.uesho.kura.file.BlockId
import com.uesho.kura.file.FileManager
import com.uesho.kura.file.Page

/**
 * The log manager, which is responsible for writing log records into a log file.
 * The tail of the log is kept in a byte buffer, which is flushed to disk when needed.
 */
class LogManager(
    private val fileManager: FileManager,
    private val logFile: String
) {
    private val logPage: Page
    private var currentBlock: BlockId
    private var latestLSN: Int = 0
    private var lastSavedLSN: Int = 0

    init {
        val bytes = ByteArray(fileManager.blockSize())
        logPage = Page(bytes)
        val logSize = fileManager.length(logFile)

        currentBlock = if (logSize == 0) {
            appendNewBlock()
        } else {
            BlockId(logFile, logSize - 1).also { block ->
                fileManager.read(block, logPage)
            }
        }
    }

    /**
     * Ensures that the log record corresponding to the specified LSN has been written to disk.
     * All earlier log records will also be written to disk.
     *
     * @param lsn the LSN of a log record
     */
    fun flush(lsn: Int) {
        if (lsn >= lastSavedLSN) {
            flush()
        }
    }

    /**
     * Appends a log record to the log buffer.
     * The record consists of an arbitrary array of bytes.
     * Log records are written right to left in the buffer.
     * The size of the record is written before the bytes.
     * The beginning of the buffer contains the location of the last-written record (the "boundary").
     * Storing the records backwards makes it easy to read them in reverse order.
     *
     * @param logRecord a byte buffer containing the bytes
     * @return the LSN of the final value
     */
    @Synchronized
    fun append(logRecord: ByteArray): Int {
        val boundary = logPage.getInt(0)
        val recordSize = logRecord.size
        val bytesNeeded = recordSize + Int.SIZE_BYTES

        if (boundary - bytesNeeded < Int.SIZE_BYTES) {
            flush()
            currentBlock = appendNewBlock()
            val newBoundary = logPage.getInt(0)
            val recordPos = newBoundary - bytesNeeded
            logPage.setBytes(recordPos, logRecord)
            logPage.setInt(0, recordPos)
        } else {
            val recordPos = boundary - bytesNeeded
            logPage.setBytes(recordPos, logRecord)
            logPage.setInt(0, recordPos)
        }

        latestLSN += 1
        return latestLSN
    }

    fun iterator(): Iterator<ByteArray> {
        flush()
        return LogIterator(fileManager, currentBlock)
    }

    /**
     * Initialize the byte buffer and append it to the log file.
     */
    private fun appendNewBlock(): BlockId {
        return fileManager.append(logFile).also { block ->
            logPage.setInt(0, fileManager.blockSize())
            fileManager.write(block, logPage)
        }
    }

    /**
     * Write the buffer to the log file.
     */
    private fun flush() {
        fileManager.write(currentBlock, logPage)
        lastSavedLSN = latestLSN
    }
}